src/Kernel/RectorKernel.php

```diff
@@ -15,51 +15,74 @@
+use Webmozart\Assert\Assert;
 
 final class RectorKernel
 {
     private readonly ConfigureCallValuesCollector $configureCallValuesCollector;
 
     private ContainerInterface|null $container = null;
 
+    /**
+     * @var array<string, ContainerInterface>
+     */
+    private static array $kernelsByHash = [];
+
     public function __construct()
     {
         $this->configureCallValuesCollector = new ConfigureCallValuesCollector();
     }
 
     /**
      * @api used in tests
+     *
+     * @param string[] $configFiles
      */
-    public function create(): ContainerInterface
+    public function createBuilder(array $configFiles = []): ContainerBuilder
     {
-        return $this->createFromConfigs([]);
+        return $this->buildContainer($configFiles);
     }
 
     /**
      * @param string[] $configFiles
      */
-    public function createFromConfigs(array $configFiles): ContainerBuilder
+    public function createFromConfigs(array $configFiles): ContainerInterface
     {
-        $defaultConfigFiles = $this->createDefaultConfigFiles();
-        $configFiles = array_merge($defaultConfigFiles, $configFiles);
-
-        $compilerPasses = $this->createCompilerPasses();
-
-        $configureCallMergingLoaderFactory = new ConfigureCallMergingLoaderFactory($this->configureCallValuesCollector);
-        $containerBuilderFactory = new ContainerBuilderFactory($configureCallMergingLoaderFactory);
-
-        $containerBuilder = $containerBuilderFactory->create($configFiles, $compilerPasses);
+        if ($configFiles === []) {
+            return $this->buildContainer([]);
+        }
 
-        // @see https://symfony.com/blog/new-in-symfony-4-4-dependency-injection-improvements-part-1
-        $containerBuilder->setParameter('container.dumper.inline_factories', true);
-        // to fix reincluding files again
-        $containerBuilder->setParameter('container.dumper.inline_class_loader', false);
+        $hash = $this->createConfigsHash($configFiles);
 
-        $containerBuilder->compile();
+        if (isset(self::$kernelsByHash[$hash])) {
+            return self::$kernelsByHash[$hash];
+        }
 
-        $this->container = $containerBuilder;
+        $className = 'RectorKernel'.$hash;
+        $file = sys_get_temp_dir() .'/rector/kernel-v3-'.$hash.'.php';
+
+        if (file_exists($file)) {
+            require_once $file;
+            $className = '\\'.__NAMESPACE__ .'\\'. $className;
+            $container = new $className();
+            if (!$container instanceof  ContainerInterface) {
+                throw new ShouldNotHappenException();
+            }
+        } else {
+            $container = $this->buildContainer($configFiles);
+
+            $dumper = new PhpDumper($container);
+
+            @mkdir(dirname($file), 0777, true);
+            file_put_contents($file, $dumper->dump([
+                'class' => $className,
+                'namespace' => __NAMESPACE__
+            ]));
+        }
```

Comment:
inspired by https://symfony.com/doc/current/components/dependency_injection/compilation.html#dumping-the-configuration-for-performance

Reply:
Could you link it in the method comment as well? Just so we know how to handle future upgrades or debug it